# Workgroup 5

# Group 6

Members:
1. Eljaer Eusebio
2. Claudia Vivas
3. Luis Sandoval
4. Andre Tapia

### Bootstraping

In [1]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split,cross_val_score
from sklearn.metrics import mean_squared_error

import matplotlib.pyplot as plt
import seaborn as sns
import pyreadr

In [2]:
## loading the data
data= pd.read_csv("../data/penn_jae.dat" , sep='\s', engine='python')
data= data[ (data['tg'] == 4) | (data['tg'] == 0) ]

In [3]:
data

abdt  tg  inuidur1  inuidur2  female  black  hispanic  othrace  dep  \
0      10824   0        18        18       0      0         0        0    2   
3      10824   0         1         1       0      0         0        0    0   
4      10747   0        27        27       0      0         0        0    0   
11     10607   4         9         9       0      0         0        0    0   
12     10831   0        27        27       0      0         0        0    1   
...      ...  ..       ...       ...     ...    ...       ...      ...  ...   
13904  10628   4        10        10       0      0         1        0    0   
13905  10523   4         4         4       0      0         1        0    2   
13906  10558   0         9         9       0      0         0        0    2   
13910  10817   4         4         4       0      0         0        0    0   
13911  10691   0        27        27       0      0         0        0    0   

       q1  ...  q5  q6  recall  agelt35  agegt54  durable  nondurable  lusd  \
0       0  ...   0   0       0        0        0        0           0     1   
3       0  ...   0   0       0        0        0        0           1     0   
4       0  ...   0   0       0        0        0        0           1     0   
11      0  ...   0   0       1        0        0        0           0     0   
12      0  ...   0   0       0        1        1        0           1     0   
...    ..  ...  ..  ..     ...      ...      ...      ...         ...   ...   
13904   0  ...   0   0       1        0        0        0           0     0   
13905   0  ...   0   0       0        0        0        0           0     0   
13906   0  ...   0   0       1        0        0        0           1     0   
13910   0  ...   0   0       0        1        0        0           0     0   
13911   0  ...   0   0       0        1        1        0           1     0   

       husd  muld  
0         0   NaN  
3         0   NaN  
4         0   NaN  
11        1   NaN  
12        0   NaN  
...     ...   ...  
13904     1   NaN  
13905     1   NaN  
13906     0   NaN  
13910     0   NaN  
13911     0   NaN  

[5099 rows x 24 columns]

In [4]:
print(data.shape)

(5099, 24)


In [5]:
# Creating the new variable
data["log(inuidur1)"] = np.log(data["inuidur1"])

In [6]:
data

abdt  tg  inuidur1  inuidur2  female  black  hispanic  othrace  dep  \
0      10824   0        18        18       0      0         0        0    2   
3      10824   0         1         1       0      0         0        0    0   
4      10747   0        27        27       0      0         0        0    0   
11     10607   4         9         9       0      0         0        0    0   
12     10831   0        27        27       0      0         0        0    1   
...      ...  ..       ...       ...     ...    ...       ...      ...  ...   
13904  10628   4        10        10       0      0         1        0    0   
13905  10523   4         4         4       0      0         1        0    2   
13906  10558   0         9         9       0      0         0        0    2   
13910  10817   4         4         4       0      0         0        0    0   
13911  10691   0        27        27       0      0         0        0    0   

       q1  ...  q6  recall  agelt35  agegt54  durable  nondurable  lusd  husd  \
0       0  ...   0       0        0        0        0           0     1     0   
3       0  ...   0       0        0        0        0           1     0     0   
4       0  ...   0       0        0        0        0           1     0     0   
11      0  ...   0       1        0        0        0           0     0     1   
12      0  ...   0       0        1        1        0           1     0     0   
...    ..  ...  ..     ...      ...      ...      ...         ...   ...   ...   
13904   0  ...   0       1        0        0        0           0     0     1   
13905   0  ...   0       0        0        0        0           0     0     1   
13906   0  ...   0       1        0        0        0           1     0     0   
13910   0  ...   0       0        1        0        0           0     0     0   
13911   0  ...   0       0        1        1        0           1     0     0   

       muld  log(inuidur1)  
0       NaN       2.890372  
3       NaN       0.000000  
4       NaN       3.295837  
11      NaN       2.197225  
12      NaN       3.295837  
...     ...            ...  
13904   NaN       2.302585  
13905   NaN       1.386294  
13906   NaN       2.197225  
13910   NaN       1.386294  
13911   NaN       3.295837  

[5099 rows x 25 columns]

In [7]:
# Creating a new variable, group treatment 4
data['T4'] = (data[['tg']]==4).astype(int)

In [8]:
data

abdt  tg  inuidur1  inuidur2  female  black  hispanic  othrace  dep  \
0      10824   0        18        18       0      0         0        0    2   
3      10824   0         1         1       0      0         0        0    0   
4      10747   0        27        27       0      0         0        0    0   
11     10607   4         9         9       0      0         0        0    0   
12     10831   0        27        27       0      0         0        0    1   
...      ...  ..       ...       ...     ...    ...       ...      ...  ...   
13904  10628   4        10        10       0      0         1        0    0   
13905  10523   4         4         4       0      0         1        0    2   
13906  10558   0         9         9       0      0         0        0    2   
13910  10817   4         4         4       0      0         0        0    0   
13911  10691   0        27        27       0      0         0        0    0   

       q1  ...  recall  agelt35  agegt54  durable  nondurable  lusd  husd  \
0       0  ...       0        0        0        0           0     1     0   
3       0  ...       0        0        0        0           1     0     0   
4       0  ...       0        0        0        0           1     0     0   
11      0  ...       1        0        0        0           0     0     1   
12      0  ...       0        1        1        0           1     0     0   
...    ..  ...     ...      ...      ...      ...         ...   ...   ...   
13904   0  ...       1        0        0        0           0     0     1   
13905   0  ...       0        0        0        0           0     0     1   
13906   0  ...       1        0        0        0           1     0     0   
13910   0  ...       0        1        0        0           0     0     0   
13911   0  ...       0        1        1        0           1     0     0   

       muld  log(inuidur1)  T4  
0       NaN       2.890372   0  
3       NaN       0.000000   0  
4       NaN       3.295837   0  
11      NaN       2.197225   1  
12      NaN       3.295837   0  
...     ...            ...  ..  
13904   NaN       2.302585   1  
13905   NaN       1.386294   1  
13906   NaN       2.197225   0  
13910   NaN       1.386294   1  
13911   NaN       3.295837   0  

[5099 rows x 26 columns]

In [9]:
data.drop(columns='muld', inplace=True)

In [10]:
data=data.reset_index(drop=True)

In [11]:
data

abdt  tg  inuidur1  inuidur2  female  black  hispanic  othrace  dep  \
0     10824   0        18        18       0      0         0        0    2   
1     10824   0         1         1       0      0         0        0    0   
2     10747   0        27        27       0      0         0        0    0   
3     10607   4         9         9       0      0         0        0    0   
4     10831   0        27        27       0      0         0        0    1   
...     ...  ..       ...       ...     ...    ...       ...      ...  ...   
5094  10628   4        10        10       0      0         1        0    0   
5095  10523   4         4         4       0      0         1        0    2   
5096  10558   0         9         9       0      0         0        0    2   
5097  10817   4         4         4       0      0         0        0    0   
5098  10691   0        27        27       0      0         0        0    0   

      q1  ...  q6  recall  agelt35  agegt54  durable  nondurable  lusd  husd  \
0      0  ...   0       0        0        0        0           0     1     0   
1      0  ...   0       0        0        0        0           1     0     0   
2      0  ...   0       0        0        0        0           1     0     0   
3      0  ...   0       1        0        0        0           0     0     1   
4      0  ...   0       0        1        1        0           1     0     0   
...   ..  ...  ..     ...      ...      ...      ...         ...   ...   ...   
5094   0  ...   0       1        0        0        0           0     0     1   
5095   0  ...   0       0        0        0        0           0     0     1   
5096   0  ...   0       1        0        0        0           1     0     0   
5097   0  ...   0       0        1        0        0           0     0     0   
5098   0  ...   0       0        1        1        0           1     0     0   

      log(inuidur1)  T4  
0          2.890372   0  
1          0.000000   0  
2          3.295837   0  
3          2.197225   1  
4          3.295837   0  
...             ...  ..  
5094       2.302585   1  
5095       1.386294   1  
5096       2.197225   0  
5097       1.386294   1  
5098       3.295837   0  

[5099 rows x 25 columns]

In [12]:
# equivalent to sample function in lab
def get_indices(data,num_samples):
    return  np.random.choice(data.index, num_samples, replace=True)

In [13]:
# Creating a function that can get estimations from the equation

def get_estimates(data,index):
    
    #Working with the equation log(inuidur1)~T4+ (female+black+othrace+factor(dep)+q2+q3+q4+q5+q6+agelt35+agegt54+durable+lusd+husd)
    
    X = data[['T4','female','black','othrace','dep','q2','q3','q4','q5','q6','agelt35','agegt54','durable','lusd','husd']].loc[index]#Independents variable
    y = data['log(inuidur1)'].loc[index] #Dependent variable
    
    #Making the regressions
    lr = LinearRegression()
    lr.fit(X,y) #Fitting the regression
    intercept = lr.intercept_
    coef = lr.coef_
    return [intercept,coef]

get_estimates(data,np.arange(0,5099))

#Creating a function that can get the coefficients and intercept through bootstraping method
def boot(data,func,R):
    intercept = []#creating an empty list
    coeff_1 = []#creating an empty list
    coeff_2 = []#creating an empty list
    coeff_3 = []#creating an empty list
    for i in range(R):
        intercept.append(func(data,get_indices(data,100))[0])#Fill in the lists
        coeff_1.append(func(data,get_indices(data,100))[1][0]) #Fill in the lists
        coeff_2.append(func(data,get_indices(data,100))[1][1])#Fill in the lists
        coeff_3.append(func(data,get_indices(data,100))[1][2])    #Fill in the lists
    
    #Getting the coefficients and intercept
    intercept_statistics = {'estimated_value':np.mean(intercept),'std_error':np.std(intercept)}   
    coeff_1_statistices = {'estimated_value':np.mean(coeff_1),'std_error':np.std(coeff_1)}   
    coeff_2_statistices = {'estimated_value':np.mean(coeff_2),'std_error':np.std(coeff_2)}
    coeff_3_statistices = {'estimated_value':np.mean(coeff_3),'std_error':np.std(coeff_3)}  
    return {'intercept':intercept_statistics,'coeff_1_statistices':coeff_1_statistices,'coeff_2_statistics':coeff_2_statistices,'coeff_3_statistics':coeff_3_statistices}

results = boot(data,get_estimates,1000)

#Let's print the results
print('Result for intercept ',results['intercept'])
print('Result for coefficient term T4',results['coeff_1_statistices'])
print('Result for coefficient term female',results['coeff_2_statistics'])
print('Result for coefficient term black',results['coeff_3_statistics'])


#lets see what the model predicts
import statsmodels.api as sm

X = data[['T4','female','black','othrace','dep','q2','q3','q4','q5','q6','agelt35','agegt54','durable','lusd','husd']]
y = data['log(inuidur1)']

X = sm.add_constant(X)
results = sm.OLS(y,X).fit()
print(results.summary())

Result for intercept  {'estimated_value': 1.7641722837034424, 'std_error': 0.3729912290101295}
Result for coefficient term T4 {'estimated_value': -0.08405992901598684, 'std_error': 0.27161537379828166}
Result for coefficient term female {'estimated_value': 0.13913457242206478, 'std_error': 0.2713836519703296}
Result for coefficient term black {'estimated_value': -0.308708296517319, 'std_error': 0.46610981854298383}
                            OLS Regression Results                            
Dep. Variable:          log(inuidur1)   R-squared:                       0.038
Model:                            OLS   Adj. R-squared:                  0.035
Method:                 Least Squares   F-statistic:                     13.34
Date:                Mon, 08 Nov 2021   Prob (F-statistic):           1.06e-33
Time:                        19:04:57   Log-Likelihood:                -8128.2
No. Observations:                5099   AIC:                         1.629e+04
Df Residuals:               

In [18]:
table = np.zeros( (1, 2) )

table[0,0] = "-0.08405992901598684"
table[0,1] = "0.27161537379828166"

table_pandas = pd.DataFrame( table, columns = [ "Estimator","Std. Error"]) 
table_pandas.index = ["T4"]
table_pandas

Estimator  Std. Error
T4   -0.08406    0.271615

In [30]:
table_1 = np.zeros( (1, 2) )

table_1[0,0] = "0.13913457242206478"
table_1[0,1] = "0.2713836519703296"

table_1_pandas = pd.DataFrame( table_1, columns = [ "Estimator","Std. Error"]) 
table_1_pandas.index = ["Female"]
table_1_pandas

Estimator  Std. Error
Female   0.139135    0.271384

In [31]:
table_2 = table_pandas.append(table_1_pandas)
table_2

Estimator  Std. Error
T4      -0.084060    0.271615
Female   0.139135    0.271384

In [32]:
table_3 = np.zeros( (1, 2) )

table_3[0,0] = "-0.308708296517319"
table_3[0,1] = "0.46610981854298383"

table_3_pandas = pd.DataFrame( table_3, columns = [ "Estimator","Std. Error"]) 
table_3_pandas.index = ["Black"]
table_3_pandas

Estimator  Std. Error
Black  -0.308708     0.46611

In [33]:
table_4 = table_2.append(table_3_pandas)
table_4

Estimator  Std. Error
T4      -0.084060    0.271615
Female   0.139135    0.271384
Black   -0.308708    0.466110

According to the bootstraping method, in the fourth treated group, says that is $8.4\%$ less likely to be unemployed for one more week compared with groups who dont belong to this fourth group. Its standar error is $0.271615$.

Aditionally, women is $13.9\%$ more likely to be unemployed for one more week compared with men. Its standar error is $0.271384$.

Finally, afro american people is $30.8708\%$ less likely to be unemployed for one more week compared with hispanic or white people. Its standar error is $0.466110$.

In [34]:
# Creating a function that can get estimations from the equation

def get_estimates(data,index):
    
    #Working with the equation log(inuidur1)~T4+ female+black
    X = data[['T4','female','black']].loc[index]#Independent variables
    y = data['log(inuidur1)'].loc[index]#Dependent variable
    
    #Making the regressions
    lr = LinearRegression()
    lr.fit(X,y)#Fitting the regression
    intercept = lr.intercept_
    coef = lr.coef_
    return [intercept,coef]

get_estimates(data,np.arange(0,5099))

#Creating a function that can get the coefficients and intercept through bootstraping method
def boot(data,func,R):
    intercept = []#Creating an empty list
    coeff_1 = []#Creating an empty list
    coeff_2 = []#Creating an empty list
    coeff_3 = []#Creating an empty list
    for i in range(R):
        intercept.append(func(data,get_indices(data,100))[0])#Fill in the lists
        coeff_1.append(func(data,get_indices(data,100))[1][0]) #Fill in the lists
        coeff_2.append(func(data,get_indices(data,100))[1][1])#Fill in the lists
        coeff_3.append(func(data,get_indices(data,100))[1][2])#Fill in the lists
    
    #Getting the coefficients and intercept
    intercept_statistics = {'estimated_value':np.mean(intercept),'std_error':np.std(intercept)}   
    coeff_1_statistices = {'estimated_value':np.mean(coeff_1),'std_error':np.std(coeff_1)}   
    coeff_2_statistices = {'estimated_value':np.mean(coeff_2),'std_error':np.std(coeff_2)}
    coeff_3_statistices = {'estimated_value':np.mean(coeff_3),'std_error':np.std(coeff_3)}  
    return {'intercept':intercept_statistics,'coeff_1_statistices':coeff_1_statistices,'coeff_2_statistics':coeff_2_statistices,'coeff_3_statistics':coeff_3_statistices}

results = boot(data,get_estimates,1000)

#Printing the results
print('Result for intercept ',results['intercept'])
print('Result for coefficient term T4',results['coeff_1_statistices'])
print('Result for coefficient term female',results['coeff_2_statistics'])
print('Result for coefficient term black',results['coeff_3_statistics'])


# for lets see what the model predicts
import statsmodels.api as sm

X = data[['T4','female','black']]
y = data['log(inuidur1)']

X = sm.add_constant(X)
results = sm.OLS(y,X).fit()
print(results.summary())

Result for intercept  {'estimated_value': 2.0429780283701753, 'std_error': 0.18793138624012345}
Result for coefficient term T4 {'estimated_value': -0.09158307570549516, 'std_error': 0.2552080957369051}
Result for coefficient term female {'estimated_value': 0.1152527713553612, 'std_error': 0.25184301249082686}
Result for coefficient term black {'estimated_value': -0.28891798816365266, 'std_error': 0.45238839933813224}
                            OLS Regression Results                            
Dep. Variable:          log(inuidur1)   R-squared:                       0.009
Model:                            OLS   Adj. R-squared:                  0.008
Method:                 Least Squares   F-statistic:                     14.79
Date:                Mon, 08 Nov 2021   Prob (F-statistic):           1.38e-09
Time:                        19:56:20   Log-Likelihood:                -8204.5
No. Observations:                5099   AIC:                         1.642e+04
Df Residuals:             

In [40]:
table_5 = np.zeros( (1, 2) )

table_5[0,0] = "-0.09158307570549516"
table_5[0,1] = "0.2552080957369051"

table_5_pandas = pd.DataFrame( table_5, columns = [ "Estimator","Std. Error"]) 
table_5_pandas.index = ["T4"]
table_5_pandas

Estimator  Std. Error
T4  -0.091583    0.255208

In [41]:
table_6 = np.zeros( (1, 2) )

table_6[0,0] = "0.1152527713553612"
table_6[0,1] = "0.25184301249082686"

table_6_pandas = pd.DataFrame( table_6, columns = [ "Estimator","Std. Error"]) 
table_6_pandas.index = ["Female"]
table_6_pandas

Estimator  Std. Error
Female   0.115253    0.251843

In [42]:
table_7 = table_5_pandas.append(table_6_pandas)
table_7

Estimator  Std. Error
T4      -0.091583    0.255208
Female   0.115253    0.251843

In [43]:
table_8 = np.zeros( (1, 2) )

table_8[0,0] = "-0.28891798816365266"
table_8[0,1] = "0.45238839933813224"

table_8_pandas = pd.DataFrame( table_8, columns = [ "Estimator","Std. Error"]) 
table_8_pandas.index = ["Black"]
table_8_pandas

Estimator  Std. Error
Black  -0.288918    0.452388

In [44]:
table_9 = table_7.append(table_8_pandas)
table_9

Estimator  Std. Error
T4      -0.091583    0.255208
Female   0.115253    0.251843
Black   -0.288918    0.452388

## Interpretations

- According to the bootstraping method, in the fourth treated group, says that is $9.1\%$ less likely to be unemployed for one more week compared with groups who dont belong to this fourth group. Its standar error is $0.255208$.

- Aditionally, women are $11.5253\%$ more likely to be unemployed for one more week compared with men. Its standar error is $0.251843$.

- Finally, afro american people are $28.89\%$ less likely to be unemployed for one more week compared with hispanic or white people. Its standar error is $0.452388$.



### Comparations
- In comparison, the standard errors of the second estimation (with all regressors) for the three variables (T4, female, and black) are smaller than the first estimation (with only the three variables T4, female, and black).